lib/letsencrypt.js

```diff
@@ -10,137 +10,126 @@
+                        log.info(`Certificate for ${params.subject} is valid till ${new Date(params.renewAt).toISOString()}`);
+                        adapter && adapter.sendToHost(adapter.host, 'certsUpdated', {validTill: params.renewAt, domain: params.subject, instance: adapter.namespace});
                     }
-                    severityFn(details);
-                }
-            };
-
-            if (settings.leStaging) {
-                log.debug('Greenlock staging enabled');
-                glOpts.staging = true;
+                    break;
             }
-
-            const gl = require('greenlock').create(glOpts);
-            gl.manager.defaults({
-                agreeToTerms: true,
-                subscriberEmail: leSettings.email
-            }).then(() => {
-                // Add sites.
-                // TODO: something seems fishy with the timing of this because on the first run
-                // a warning about no sites being configured is found, even though things go on to
-                // work perfectly.
-                for (const domain of leSettings.domains) {
-                    log.debug(`Adding to Greenlock: ${domain}`);
-                    gl.add({ subject: domain, altnames: [domain] });
-                }
-
-                require('greenlock-express')
-                    .init({
-                        greenlock: gl
-                    }).ready(glx => {
-                        // The below is from greenlock-express examples/https/server.js
-
-                        // Start the challenge server if configured
-                        if (settings.leUpdate) {
-                            const httpServer = glx.httpServer();
-                            settings.lePort = parseInt(settings.lePort, 10) || 80;
-                            const bind = (!settings.bind || settings.bind === '0.0.0.0') ? undefined : settings.bind || undefined;
-                            httpServer.listen(settings.lePort, bind, () =>
-                                log.info('Challenge server is started on ' + httpServer.address()));
-                        }
-
-                        // Don't call listen here because caller will do that
-                        resolve(glx.httpsServer(null, app));
-                    });
-            }).catch(err => {
-                reject(new Error(`Failed to set Greenlock manager defaults: ${err}`));
-            });
         }
+    };
+
+    if (settings.leStaging) {
+        log.debug('Greenlock staging enabled');
+        glOpts.staging = true;
+    }
+
+    const gl = require('greenlock').create(glOpts);
+
+    // Add sites.
+    for (const domain of leSettings.domains) {
+        log.debug(`Adding to Greenlock: ${domain}`);
+        gl.add({ subject: domain, altnames: [domain] });
+    }
+
+    await gl.manager.defaults({
+        agreeToTerms: true,
+        subscriberEmail: leSettings.email
+    });
+
+    return await new Promise((resolve, reject) => {
```

Comment:
I think a normal "return new Promise" is also fine here. the await is from outside in anycase

Reply:
Well, you notice that I made the `createServerAsync` function `async`. If one returns a `new Promise` from within an `async` function one is actually returning a Promise wrapped in a Promise.

It was necessary to create this Promise here because the Greenlock module has no `async` functions and uses callbacks.

Reply:
From my knowlede JS detects that and if you return a promise it will not be wrapped again

Reply:
BTW, yes, we could `await adapter.getPortAsync` outside of that Promise block but I do not like that as there would have to be two checks for `if (settings.leUpdate)...`

Reply:
> From my knowlede JS detects that and if you return a promise it will not be wrapped again

Hmmmm... this could be true, but I see you merged it, so, erm, we worry about it later ;)